### ISSUE TITLE
[RA] Enable naval units in skirmish

### ISSUE THREAD

**Issue by TobiasKarnat** [TheAssemblyArmada/Vanilla-Conquer#225]
This would enable building of naval vessels in RA skrimish

Updated to check for ship yard or sub pen as well in Fire_Sale()

```
diff -urN a/redalert/house.cpp b/redalert/house.cpp
--- a/redalert/house.cpp	2021-01-15 14:07:59.350866822 +0100
+++ b/redalert/house.cpp	2021-01-15 14:15:22.647845176 +0100
@@ -5257,7 +5257,7 @@
     */
     if (State != STATE_ATTACKED && CurBuildings
         && !(ActiveBScan
-             & (STRUCTF_TENT | STRUCTF_BARRACKS | STRUCTF_CONST | STRUCTF_AIRSTRIP | STRUCTF_WEAP | STRUCTF_HELIPAD))) {
+             & (STRUCTF_TENT | STRUCTF_BARRACKS | STRUCTF_CONST | STRUCTF_AIRSTRIP | STRUCTF_WEAP | STRUCTF_HELIPAD | STRUCTF_SHIP_YARD | STRUCTF_SUB_PEN))) {
         return (URGENCY_CRITICAL);
     }
     return (URGENCY_NONE);
@@ -5348,6 +5348,24 @@
             }
         }
     }
+    for (index = 0; index < Vessels.Count(); index++) {
+		VesselClass* v = Vessels.Ptr(index);
+
+		if (v != NULL && !v->IsInLimbo && v->House == this && v->Strength > 0) {
+			if (!shuffle && v->Is_Weapon_Equipped() && (forced || Percent_Chance(75))) {
+				v->Assign_Mission(MISSION_HUNT);
+			} else {
+
+				/*
+				**	If this vessel is guarding the base, then cause it to shuffle
+				**	location instead.
+				*/
+				if (Percent_Chance(20) && v->Mission == MISSION_GUARD_AREA && Which_Zone(v) != ZONE_NONE) {
+					v->ArchiveTarget = ::As_Target(Where_To_Go(v));
+				}
+			}
+		}
+	}
     for (index = 0; index < Units.Count(); index++) {
         UnitClass* u = Units.Ptr(index);
 
@@ -5505,8 +5523,6 @@
         StructType Structure;
         UrgencyType Urgency;
     } _types[] = {{STRUCT_CHRONOSPHERE, URGENCY_LOW},
-                  {STRUCT_SHIP_YARD, URGENCY_LOW},
-                  {STRUCT_SUB_PEN, URGENCY_LOW},
                   {STRUCT_ADVANCED_TECH, URGENCY_LOW},
                   {STRUCT_FORWARD_COM, URGENCY_LOW},
                   {STRUCT_SOVIET_TECH, URGENCY_LOW},
@@ -5559,8 +5575,6 @@
         StructType Structure;
         UrgencyType Urgency;
     } _types[] = {{STRUCT_CHRONOSPHERE, URGENCY_LOW},
-                  {STRUCT_SHIP_YARD, URGENCY_LOW},
-                  {STRUCT_SUB_PEN, URGENCY_LOW},
                   {STRUCT_ADVANCED_TECH, URGENCY_LOW},
                   {STRUCT_FORWARD_COM, URGENCY_LOW},
                   {STRUCT_SOVIET_TECH, URGENCY_LOW},
@@ -5841,6 +5855,29 @@
             }
         }
 
+		/*
+		**	A source of naval vessels is always needed, but only if there will
+		**	be sufficient money to build vessels.
+		*/
+		current = BQuantity[STRUCT_SHIP_YARD] + BQuantity[STRUCT_SUB_PEN];
+		if (current < 2 && (money > 650 || hasincome)) {
+			b = &BuildingTypeClass::As_Reference(STRUCT_SHIP_YARD);
+			if (Can_Build(b, ActLike) && (b->Cost_Of() < money || hasincome)) {
+				choiceptr = BuildChoice.Alloc();
+				if (choiceptr != NULL) {
+					*choiceptr = BuildChoiceClass(current > 0 ? URGENCY_LOW : URGENCY_MEDIUM, b->Type);
+				}
+			} else {
+				b = &BuildingTypeClass::As_Reference(STRUCT_SUB_PEN);
+				if (Can_Build(b, ActLike) && (b->Cost_Of() < money || hasincome)) {
+					choiceptr = BuildChoice.Alloc();
+					if (choiceptr != NULL) {
+						*choiceptr = BuildChoiceClass(current > 0 ? URGENCY_LOW : URGENCY_MEDIUM, b->Type);
+					}
+				}
+			}
+		}
+
         /*
         **	Always build up some base defense.
         */
@@ -6299,9 +6336,32 @@
         }
     }
 
-    if (IsBaseBuilding) {
-        BuildVessel = VESSEL_NONE;
-    }
+	if (IsBaseBuilding) {
+
+		int counter[VESSEL_COUNT];
+		int total = 0;
+		VesselType index;
+		for (index = VESSEL_FIRST; index < VESSEL_COUNT; index++) {
+			VesselTypeClass const* vtype = &VesselTypeClass::As_Reference(index);
+			if (Can_Build(vtype, ActLike) && vtype->PrimaryWeapon != NULL) {
+				counter[index] = 1;
+            } else {
+                counter[index] = 0;
+            }
+            total += counter[index];
+		}
+
+		if (total > 0) {
+			int choice = Random_Pick(0, total - 1);
+			for (index = VESSEL_FIRST; index < VESSEL_COUNT; index++) {
+				if (choice < counter[index]) {
+					BuildVessel = index;
+					break;
+				}
+				choice -= counter[index];
+			}
+		}
+	}
 
     return (TICKS_PER_SECOND);
 }
```

**Comment by mvdhout1992** [comment]
Does the AI actually use the naval units? And what about cash management? The AI runs out of cash really quick so it would be nice to have the AI cheat with the amount of credits it gets from harvesting ore. I did this in my ASM patch and it fixed most of the AI credit management issues (except the AI ore truck getting stuck).

**Comment by TobiasKarnat** [comment]
Yes, the AI does use the units against you. Try it ;-D

Additionally the use of the naval transport unit like the AI already does for the APC would be nice to have.

Yeah cash is very limited, even more in TD skirmish. Would make sense to add.

**Comment by TobiasKarnat** [comment]
This will shuffle the destination for build ships and submarines (unblocks building of multiple units).
```
--- a/redalert/building.cpp	2021-01-22 12:50:09.019964241 +0100
+++ b/redalert/building.cpp	2021-01-22 13:28:00.273877139 +0100
@@ -2074,6 +2074,7 @@
             cell = Find_Exit_Cell(base);
             if (cell != 0 && base->Unlimbo(Cell_Coord(cell), Direction(Cell_Coord(cell)))) {
                 base->Assign_Mission(MISSION_GUARD);
+                base->Assign_Destination(::As_Target(base->Nearby_Location(this)));
                 ScenarioInit--;
                 return (2);
             }
```
